akka-stream/src/main/scala/akka/stream/impl/io/IOSinks.scala

```diff
@@ -66,3 +72,90 @@ private[akka] final class OutputStreamSink(createOutput: () ⇒ OutputStream, va
   override def withAttributes(attr: Attributes): Module =
     new OutputStreamSink(createOutput, attr, amendShape(attr))
 }
+
+/**
+ * INTERNAL API
+ * Creates simple synchronous (Java 6 compatible) Sink which provide all incoming element on read request via
+ * InputStream.
+ */
+private[akka] final class InputStreamSink(timeout: FiniteDuration, val attributes: Attributes, shape: SinkShape[ByteString])
+  extends SinkModule[ByteString, (InputStream, Future[Long])](shape) {
+
+  override def create(context: MaterializationContext) = {
+    val mat = ActorMaterializer.downcast(context.materializer)
+    val settings = mat.effectiveSettings(context.effectiveAttributes)
+    val bytesWrittenPromise = Promise[Long]()
+
+    val props = InputStreamSubscriber.props(bytesWrittenPromise, settings.maxInputBufferSize)
+    val dispatcher = IOSettings.blockingIoDispatcher(context)
+
+    val ref = mat.actorOf(context, props.withDispatcher(dispatcher))
+    (akka.stream.actor.ActorSubscriber[ByteString](ref),
+      (new InputStreamAdapter(ref, timeout).asInstanceOf[InputStream], bytesWrittenPromise.future))
+  }
+
+  override protected def newInstance(shape: SinkShape[ByteString]): SinkModule[ByteString, (InputStream, Future[Long])] =
+    new InputStreamSink(timeout, attributes, shape)
+
+  override def withAttributes(attr: Attributes): Module =
+    new InputStreamSink(timeout, attr, amendShape(attr))
+}
+
+/**
+ * INTERNAL API
+ * For Java API usages (returns Java API Pair)
+ */
+private[akka] final class JavaInputStreamSink(timeout: FiniteDuration, val attributes: Attributes, shape: SinkShape[ByteString])
+  extends SinkModule[ByteString, japi.Pair[InputStream, Future[JLong]]](shape) {
+  import akka.dispatch.ExecutionContexts.{ sameThreadExecutionContext ⇒ ec }
+
+  val sink = new InputStreamSink(timeout, attributes, shape)
+
+  override def create(context: MaterializationContext) = {
+    val result = sink.create(context)
+    (result._1, new japi.Pair(result._2._1, result._2._2.map(long2Long)(ec)))
```

Comment:
this is a classical case of where the `akka.dispatch.ExecutionContexts.sameThreadExecutionContext` is appropriate